core/src/main/scala/org/apache/spark/sql/delta/RowId.scala

```diff
@@ -87,4 +90,120 @@ object RowId {
+          case a => a
+        }
+      } else {
+        val watermarkToCommit = newHighWatermark
+        newHighWatermark = oldHighWatermark // reset high watermark to make hasNext return false
+        RowIdHighWaterMark(watermarkToCommit, preservedRowIds = false)
+      }
+    }
+  }
+
+  /**
+   * Assigns fresh row IDs to all AddFiles inside `actions` that do not have row IDs yet.
+   */
+  private[delta] def assignFreshRowIds(
+      spark: SparkSession,
+      protocol: Protocol,
+      snapshot: Snapshot,
+      actions: Iterator[Action]): Iterator[Action] = {
+    if (rowIdsAllowed(spark) && rowIdsSupported(protocol)) {
+      val highWatermark = extractHighWatermark(spark, snapshot).map(_.highWaterMark)
+      new RowIdAssigningIterator(actions, highWatermark)
+    } else {
+      actions
+    }
+  }
+
+  /**
+   * Reassigns newly assigned Row IDs (i.e. Row IDs above `readHighWaterMark`) of all AddFile
+   * actions in `actions`  to be above `winningHighWaterMarkOpt` (if it is defined), and
+   * updates the high watermark in `actions`.
+   */
+  private[delta] def reassignOverlappingRowIds(
+      readHighWaterMark: Long,
+      winningHighWaterMarkOpt: Option[Long],
+      actions: Seq[Action]): Seq[Action] = {
+    if (winningHighWaterMarkOpt.isEmpty) {
+      // The winning transaction did not assign any new row IDs, so there is no need to check for
+      // duplicate row IDs.
+      return actions
+    }
+    val winningHighWaterMark = winningHighWaterMarkOpt.get
+
+    assert(winningHighWaterMark >= readHighWaterMark)
+    val watermarkDiff = winningHighWaterMark - readHighWaterMark
+    actions.map {
+      case a: AddFile =>
+        // We should only update the row IDs that were assigned by this transaction, and not the
+        // row IDs that were assigned by an earlier transaction and merely copied over to a new
+        // AddFile as part of this transaction. I.e., we should only update the base row IDs that
+        // are larger than the read high watermark.
+        a.baseRowId match {
+          case Some(baseRowId) if baseRowId > readHighWaterMark =>
+            val newBaseRowId = baseRowId + watermarkDiff
+            a.copy(baseRowId = Some(newBaseRowId))
+          case _ =>
+            a
+        }
+
+      case RowIdHighWaterMark(v, preservedRowIds) =>
+        RowIdHighWaterMark(v + watermarkDiff, preservedRowIds)
+
+      case a =>
+        a
+    }
+  }
+
+  /**
+   * Extracts the high watermark of row IDs from a snapshot.
+   */
+  private[delta] def extractHighWatermark(
+      spark: SparkSession, snapshot: Snapshot): Option[RowIdHighWaterMark] = {
+    if (rowIdsAllowed(spark)) {
+      snapshot.rowIdHighWaterMarkOpt
+    } else {
+      None
+    }
+  }
+
+  private[delta] def extractHighWatermark(
+      spark: SparkSession, actions: Seq[Action]): Option[RowIdHighWaterMark] = {
+    if (rowIdsAllowed(spark)) {
+      actions.collectFirst { case r: RowIdHighWaterMark => r }
```

Comment:
how about findLast here because
* the collect does not actually transform
* the action is likely at the very end
* if the sequence is spanning more than one commit, it would be the correct one